airbyte-integrations/connectors/source-snapchat-marketing/source_snapchat_marketing/source.py

```diff
@@ -0,0 +1,397 @@
+            records = depend_on_stream.read_records(sync_mode=SyncMode.full_refresh, stream_slice=depend_on_stream_slice)
+            depend_on_stream_ids += [{slice_key_name: record["id"]} for record in records]
+    else:
+        depend_on_stream_ids = [{slice_key_name: record["id"]} for record in depend_on_stream.read_records(sync_mode=SyncMode.full_refresh)]
+
+    if not depend_on_stream_ids:
+        return []
+
+    auxiliary_id_map[depends_on_stream.__name__] = depend_on_stream_ids
+    return depend_on_stream_ids
+
+
+class SnapchatMarketingStream(HttpStream, ABC):
+    url_base = "https://adsapi.snapchat.com/v1/"
+    primary_key = "id"
+
+    @property
+    def response_root_name(self):
+        """ Using the class name in lower to set the root node for response parsing """
+        return self.name
+
+    @property
+    def response_item_name(self):
+        """
+        Used as the second level node for response parsing. Usually it is the response_root_name without last char.
+        For example: response_root_name = organizations, so response_item_name = organization
+        The [:-1] removes the last char from response_root_name: organizations -> organization
+         {
+            "organizations": [
+                {
+                    "organization": {....}
+                }
+            ]
+        }
+        """
+        return self.response_root_name[:-1]
+
+    def __init__(self, start_date, **kwargs):
+        super().__init__(**kwargs)
+        self.start_date = pendulum.parse(start_date).to_rfc3339_string()
+
+    def next_page_token(self, response: requests.Response) -> Optional[Mapping[str, Any]]:
+        next_page_cursor = response.json().get("paging", False)
+        if next_page_cursor:
+            return {"cursor": dict(parse_qsl(urlparse(next_page_cursor["next_link"]).query))["cursor"]}
+
+    def request_params(
+        self, stream_state: Mapping[str, Any], stream_slice: Mapping[str, any] = None, next_page_token: Mapping[str, Any] = None
+    ) -> MutableMapping[str, Any]:
+        return next_page_token or {}
+
+    def parse_response(self, response: requests.Response, **kwargs) -> Iterable[Mapping]:
+        """Response json came like
+        {
+            "organizations": [
+                {
+                    "organization": {
+                        "id": "some uuid",
+                        "updated_at": "2020-12-15T22:35:17.819Z",
+                        "created_at": "2020-12-15T11:13:03.910Z",
+                        ... some_other_json_fields ...
+                    }
+                }
+            ]
+        }
+        So the response_root_name will be "organizations", and the response_item_name will be "organization"
+        Also, the client side filtering for incremental sync is used
+        """
+        json_response = response.json().get(self.response_root_name)
+        for resp in json_response:
+            yield resp.get(self.response_item_name)
```

Comment:
Related to my point above on `response_item_name`, I think this will silently yield None if we've got the `response_item_name` incorrect. Perhaps we could error or log a WARN message if json_response is non-empty but we're only yielding Nones as this would probably indicate we're missing data (unless the responses get packed with extra junk we don't care about)

Reply:
See my [comment above](https://github.com/airbytehq/airbyte/pull/4843#discussion_r678459158)